### DISCUSSION TITLE
Extract table without vertical or horizontal lines

### DISCUSSION THREAD

**Discussion by kucoll** [jsvine/pdfplumber#1060] [Ask for help with specific PDFs]
Hello,

First, thank you for this awesome library, it solves most of my problems.

I need your help with a few questions about borderless table extraction, can you provide some advice and help?

# First Case debug table
![tmp6wysn5jb](https://github.com/jsvine/pdfplumber/assets/5094212/3508c160-ebce-49bb-bd58-1ca14c5a1783)

My code is below：
```
doc = pdfplumber.open(pdf_file)
page = doc.pages[0]
table_settings = {
    "snap_x_tolerance": 8
}
page.to_image(resolution=200).debug_tablefinder(table_settings).show()
```

This case is on page 1 of the pdf file.  [join.pdf](https://github.com/jsvine/pdfplumber/files/13664645/join.pdf)

1. The first borderless table on this page is not recognized, how should I adjust the code to recognize it correctly?
2. How to filter out the meaningless little border in the second table?

# Second Case debug table

![tmpdze_h9ab](https://github.com/jsvine/pdfplumber/assets/5094212/c1679d3b-05f4-4802-83e8-097a2010f2e6)

My code is below：
```
doc = pdfplumber.open(pdf_file)
page = doc.pages[1]
table_settings = {}
page.to_image(resolution=200).debug_tablefinder(table_settings).show()
```
This case is on page 2 of the pdf file.  [join.pdf](https://github.com/jsvine/pdfplumber/files/13664645/join.pdf)

I want to extract two tables from the first page, the second table is perfect, but the first one seems to need to adjust the code?

# Third  Case debug table
![tmpuyo2d_gd](https://github.com/jsvine/pdfplumber/assets/5094212/9f4f6578-0e55-437d-95c9-967a29ac5c9e)

My code is below：
```
doc = pdfplumber.open(pdf_file)
page = doc.pages[2]

vlines = [e for e in page.vertical_edges if e["height"] > 10]
extreme_xs = [
    min(e["x0"] for e in page.horizontal_edges),
    max(e["x1"] for e in page.horizontal_edges)
]
table_settings = {
    "horizontal_strategy": "text",
    "explicit_horizontal_lines": page.curves + page.edges,
    "explicit_vertical_lines": vlines + extreme_xs,
}
page.to_image(resolution = 200).debug_tablefinder(table_settings).show()
```
This case is on page 3 of the pdf file.  [join.pdf](https://github.com/jsvine/pdfplumber/files/13664645/join.pdf)

I don't know how to extract the two forms on the third page, and I don't need the text between the two forms？

# Fourth Case：
![tmp4eryu30y](https://github.com/jsvine/pdfplumber/assets/5094212/33697fa0-c8c3-4da1-83bd-ff6f2c0de15c)

My code is below：
```
doc = pdfplumber.open(pdf_file)
page = doc.pages[1]

vlines = [e for e in page.vertical_edges if e["height"] > 10]
extreme_xs = [
    min(e["x0"] for e in page.horizontal_edges),
    max(e["x1"] for e in page.horizontal_edges)
]
table_settings = {
    "vertical_strategy": "text",
    "horizontal_strategy": "text",
}
page.to_image(resolution = 200).debug_tablefinder(table_settings).show()
```
This case is on page 5 of the pdf file.  [join.pdf](https://github.com/jsvine/pdfplumber/files/13664645/join.pdf)

How should I extract two tables without the text in the middle, should I filter out the line where the text "在建开发产品" is in python code?

# Five Case：
![image](https://github.com/jsvine/pdfplumber/assets/5094212/5ff37cba-353d-4615-8468-d74455d1b089)

The pdf file of this case is in [join.pdf](https://github.com/jsvine/pdfplumber/files/13664645/join.pdf)

How can I identify the table on page 3 and page 4 as a whole table, or is there any way to identify the table data across pages?

@jsvine  Or others Can you give me some advice about the above five cases? I would be very happy if I could get your help。

**Comment by jsvine** [comment]
Hi @kucoll, and thanks for your interest in `pdfplumber`, and for providing detailed explanations, as well as the relevant PDF. 

### First Case 

> The first borderless table on this page is not recognized, how should I adjust the code to recognize it correctly?

This will be a tricky table to recognize correctly, because it contains no obvious vertical separators. You can use `"vertical_strategy": "text"`, although this will also affect the extraction of the other tables on the page.

> How to filter out the meaningless little border in the second table?

The additional borders in the second and third tables appear to be caused by "invisible" lines — lines encoded in the PDF, but in a manner that make them invisible to the human eye. I'd suggest examining the results of `page.lines` and `page.rects` to determine the unique characteristics of those lines, and then use `page.filter(...).extract_tables(...)` to filter them out before extraction. (You can find some examples of this approach in other discussions in the forum.) 

### Second Case 

> I want to extract two tables from the first page, the second table is perfect, but the first one seems to need to adjust the code?

This is a similar situation as the first table in the prior example; I'd suggest the same approach.

### Third Case / Fourth Case

This is a similar situation to the "invisible" lines discussed above; I'd suggest a similar approach (although in this case there might be some "invisible" lines you want to keep).

### Fifth Case

> How can I identify the table on page 3 and page 4 as a whole table, or is there any way to identify the table data across pages?

Unfortunately, there is no `pdfplumber` method for analyzing tables across pages. This will require you to write your own logic. There are some prior discussions that might be useful:

- https://github.com/jsvine/pdfplumber/discussions/1019
- https://github.com/jsvine/pdfplumber/discussions/768

**Comment by Eyderoe** [comment]
@kucoll  你好， 我大概看了一下，有一些小建议。提取表格不一定非要提取表格，也可以先提取words再组成一个表格(缺点就是表头不一定正常，也许要对表头单独处理)，因为这个表格很工整，缺的地方都有-来代替，同一行可以取这个词的上下边界中间值确定，一定误差范围内可以被认为是同一行。对于表格本身，我想可以通过字体大小来判断，因为表格字体好像比其他的要小一号。words提取出来["char"]["size"]可以看大小，根据这个大字号文本的分布，可以划分出表格可能分布的区间。hmm，想了下，通过判断小字中间有没有大字好像要更容易一些。这样的话，也可以解决跨页面的表格，或者判断表格第一行有多少词应该也行。

**Comment by clj55** [comment]
Hi I'm facing issues extracting the invisible tables too. I can't crop the page to specific coordinates of the table because I'm running the program on multiple PDFs where the table can appear in different positions.

https://github.com/user-attachments/files/15922643/Customer.Information.pdf

I tried it with these table_settings:
`table_settings={
    "vertical_strategy":"text",
                "text_keep_blank_chars":True,
                "horizontal_strategy":"text",
                }`
But it recognised the paragraph text as a table too

<img width="545" alt="341638091-8bef3510-9f8b-4ca3-811e-07f5dbaec5b7" src="https://github.com/jsvine/pdfplumber/assets/103747880/8a9e5c29-adc7-4430-af95-e3669ce59bc6">

I inspected the page.rects and the page.lines but 
page.rects: identifies the text underlines
page.lines: empty list

Does anyone know how to solve this?

(reposted from https://github.com/jsvine/pdfplumber/issues/123#issuecomment-2181968934)

**Reply by Eyderoe** [reply]
> Hi I'm facing issues extracting the invisible tables too. I can't crop the page to specific coordinates of the table because I'm running the program on multiple PDFs where the table can appear in different positions.
> 
> https://github.com/user-attachments/files/15922643/Customer.Information.pdf
> 
> I tried it with these table_settings: `table_settings={ "vertical_strategy":"text", "text_keep_blank_chars":True, "horizontal_strategy":"text", }` But it recognised the paragraph text as a table too
> 
> <img alt="341638091-8bef3510-9f8b-4ca3-811e-07f5dbaec5b7" width="545" src="https://private-user-images.githubusercontent.com/103747880/344074158-8a9e5c29-adc7-4430-af95-e3669ce59bc6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk1Nzg0ODYsIm5iZiI6MTcxOTU3ODE4NiwicGF0aCI6Ii8xMDM3NDc4ODAvMzQ0MDc0MTU4LThhOWU1YzI5LWFkYzctNDQzMC1hZjk1LWUzNjY5Y2U1OWJjNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyOFQxMjM2MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ODcyOTI1NjA4YjRkNTQ1ZjBhZjhmMzJhZjkzOTM1MDE1MzlmODQ4ZDA0MGYyYTcyYzZiODNlZWY1OWU2MDk0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Axvl6m3_t78l-qpGAPfxyzkqE3FutVuu_hI6lm8Y-8c">
> I inspected the page.rects and the page.lines but page.rects: identifies the text underlines page.lines: empty list
> 
> Does anyone know how to solve this?
> 
> (reposted from [#123 (comment)](https://github.com/jsvine/pdfplumber/issues/123#issuecomment-2181968934))

have more sample ?
or all the files are single page and nothing below table ?
